runtime/src/accounts_db.rs

```diff
@@ -1925,16 +1958,63 @@ impl AccountsDb {
         Self::calc_delete_dependencies(&purges_zero_lamports, &mut store_counts);
         calc_deps_time.stop();
 
-        // Only keep purges_zero_lamports where the entire history of the account in the root set
-        // can be purged. All AppendVecs for those updates are dead.
+        let should_filter_for_incremental_snapshots =
+            max_clean_root.unwrap_or(Slot::MAX) > last_full_snapshot_slot.unwrap_or(Slot::MAX);
+        assert!(
+            last_full_snapshot_slot.is_some() || !should_filter_for_incremental_snapshots,
+            "if filtering for incremental snapshots, then snapshots should be enabled",
+        );
+
         let mut purge_filter = Measure::start("purge_filter");
-        purges_zero_lamports.retain(|_pubkey, (account_infos, _ref_count)| {
-            for (_slot, account_info) in account_infos.iter() {
+        purges_zero_lamports.retain(|pubkey, (slot_account_infos, _ref_count)| {
+            // Only keep purges_zero_lamports where the entire history of the account in the root set
+            // can be purged. All AppendVecs for those updates are dead.
+            for (_slot, account_info) in slot_account_infos.iter() {
                 if store_counts.get(&account_info.store_id).unwrap().0 != 0 {
                     return false;
                 }
             }
-            true
+
+            // Exit early if not filtering more for incremental snapshots
+            if !should_filter_for_incremental_snapshots {
+                return true;
+            }
+
+            // When using incremental snapshots, do not purge zero-lamport accounts if the slot is
+            // higher than the last full snapshot slot.  This is to protect against the following
+            // scenario:
+            //
+            //      A full snapshot is taken, and it contains an account with a non-zero balance.
+            //      Later, that account's  goes to zero.  Evntually cleaning runs, and before, this
+            //      account would be cleaned up.  Finally, an incremental snapshot is taken.
+            //
+            //      Later, the incremental (and full) snapshot is used to rebuild the bank and
+            //      accounts database (e.x. if the node restarts).  The full snapshot _does_
+            //      contain the account (from above) and its balance is non-zero, however, since
+            //      the account was cleaned up in a later slot, the incremental snapshot does not
+            //      contain any info about this account, thus, the accounts database will contain
+            //      the old info from this account, which has its old non-zero balance.  Very bad!
```

Comment:
Is this issue only limited to zero lamports? What about for older roots with non-zero lamports which are cleaned up because they were superseded by updates in later slots?

Reply:
For example for an account it is updated at slot 1, slot 2 and slot 3 with non zero lamports. We take full snapshot at slot 1, accounts cleanup happens at slot 3 (slot2) can be cleaned up as changes superseded by update in slot 3. And later an incremental snapshot is taken at slot 2. Can this happen?

Reply:
Correct, the issue only applies to zero-lamport accounts. When an account has a non-zero balance, it will not be removed when clean is called, which is the crux of the issue.

This is a good question about snapshots that occur at slots lower than the max rooted slot. This is the same issue that could occur today without incremental snapshot support. Since clean only occurs in the background services which also handle snapshots, this does two things: (1) clean knows what slot it is snapshot-ing, and (2) what slot the bank/bank forks is currently on. These two can make sure that clean doesn't remove dead stores before snapshots complete.

Or at least that's my understanding of how this process works! I reserve the right to be wrong though :)

Reply:
@carllin, @CriesofCarrots  What is your take? I agree that we may have the same issue with full snap. This is probably corrected by repair_stage later? But if the RPC do a scan at that slot, would not we give wrong information?

Reply:
> For example for an account it is updated at slot 1, slot 2 and slot 3 with non zero lamports. We take full snapshot at slot 1, accounts cleanup happens at slot 3 (slot2) can be cleaned up as changes superseded by update in slot 3. And later an incremental snapshot is taken at slot 2. Can this happen

This can't currently happen because the clean is synchronous with the storage fetch in AccountsBackgroundService. This means we save off all storage entries necessary for snapshotting up to slot `X` before we clean past slot `X`. this guarantees the storage entry for slot 2 is available for the snapshot, and has not yet been removed.